SimPEG/Directives.py

```diff
@@ -730,60 +733,62 @@ class Update_lin_PreCond(InversionDirective):
     Create a Jacobi preconditioner for the linear problem
     """
     onlyOnStart = False
-    mapping = None
-    ComboObjFun = False
 
     def initialize(self):
 
-        # Check if it is a ComboObjective
-        if not isinstance(self.reg, Regularization.BaseComboRegularization):
+        # Create the pre-conditioner
+        regDiag = np.zeros_like(self.invProb.model)
 
-            # It is a Combo objective, so will have to loop
-            self.ComboObjFun = True
+        for reg in self.reg.objfcts:
+            # Check if he has wire
+            if getattr(reg.mapping, 'P', None) is None:
+                regDiag += (reg.W.T*reg.W).diagonal()
+            else:
+                # He is a snitch!
+                regDiag += reg.mapping.P.T*(reg.W.T*reg.W).diagonal()
 
-        if getattr(self, 'mapping', None) is None:
-            self.mapping = Maps.IdentityMap(nP=self.reg.mapping.nP)
 
-        if getattr(self.opt, 'approxHinv', None) is None:
+        # Deal with the linear case
+        if getattr(self.opt, 'JtJdiag', None) is None:
 
-            # Update the pre-conditioner
-            if self.ComboObjFun:
+            print("Approximated diag(JtJ) with linear operator")
 
-                reg_diag = []
-                for reg in self.reg.objfcts:
-                    reg_diag.append(self.invProb.beta*(reg.W.T*reg.W).diagonal())
+            JtJdiag = np.zeros_like(self.invProb.model)
 
-                diagA = np.sum(self.prob.G**2., axis=0) + np.hstack(reg_diag)
+            for prob, dmisfit in zip(self.prob, self.dmisfit.objfcts):
 
-            else:
-                diagA = (np.sum(self.prob.G**2., axis=0) +
-                         self.invProb.beta*(self.reg.W.T*self.reg.W).diagonal())
+                wd = dmisfit.W.diagonal()
+                for ii in range(prob.G.shape[0]):
+                    JtJdiag += prob.mapping().deriv().T*(wd[ii] * prob.G[ii, :])**2.
```

Comment:
Ok good one, I have a question about that. Here the mapping is a projection having the name 'chiMap' because it's the mag problem. But other problems will use other mapping 'sigmaMap', 'muMap' etc. We should just be able to just call 'mapping' from the outside, without having to be specific about a problem. How to do that?

Reply:
Got it, "prob.mapPair()"